compiler/rustc_mir_transform/src/fold.rs

```diff
@@ -0,0 +1,385 @@
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    if new_place == *final_place {
+        Some(StatementKind::Nop)
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Fold:
+// _2 = &mut _1;
+// _3 = &raw mut *_2;
+// Into:
+// _3 = &raw mut _1;
+fn fold_ref_addressof<'tcx, 'a>(
+    cx: &FoldContext<'tcx, 'a>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::AddressOf(mutability, second_place) = final_rvalue else { return None; };
+
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    Some(StatementKind::Assign(Box::new((*final_place, Rvalue::AddressOf(*mutability, new_place)))))
+}
+
+// Fold:
+// _2 = _1.a;
+// _3 = _2.b;
+// Into:
+// _3 = _1.a.b;
+fn fold_place_projections<'tcx, 'a>(
+    cx: &FoldContext<'tcx, 'a>,
+    temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Use(temp_operand) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    // Both the operands needs to be places (is it even possible for one to be a Constant?)
+    let first_place = temp_operand.place()?;
+    let second_place = final_operand.place()?;
+
+    // If we are assigning into a place expression, that would be something like
+    // _3.b = _2.b;
+    // Which is complicated. Just don't optimize that at all for now.
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+
+    // Derefs must come first if at all
+    // If merging these assignments would break that rule, bummer. Bail.
+    if first_place.projection.len() > 0
+        && second_place.projection.get(0) == Some(&ProjectionElem::Deref)
+    {
+        return None;
+    }
+
+    // If the temporary has a niche but the final does not, doing this optimization will destroy
+    // the niche information.
+    // This check is _extremely_ cautious, we only do this optimization if we are absolutely
+    // certain that the temporary does not have a niche.
+    // Note that if the second assignment does not add any projections, layout can't change, so we
+    // don't need this check.
+    if !second_place.projection.is_empty() {
+        let temporary_ty = temp_place.ty(cx.local_decls, cx.tcx).ty;
+        let Ok(layout) = cx.tcx.layout_of(cx.param_env.and(temporary_ty)) else {
+            return None;
+        };
+        if layout.layout.largest_niche().is_some() {
+            return None;
+        }
+    }
```

Comment:
I'm not sure I understand the issue with niches. Could you add a test exhibiting it?

Reply:
Without this check, `tests/codegen/vec-as-ptr.rs` and `tests/codegen/slice-windows-no-bounds-check.rs` fail.

Does that answer your question? I feel like those check this a bit indirectly, so there is probably a better way to answer your question but I just can't come up with anything.

Reply:
This is probably something that should be handled by codegen. IIUC, if you have a `x: (NonZeroUsize,)`, `x.0` in MIR gets the range metadata for codegen, and `x.0.0` does not. Is that so?

Reply:
I don't know.

Some quick searching says LLVM range metadata is for integer loads only. The problematic function that shows up in codegen tests is `Vec::as_ptr`, which does a `load ptr`, and the change to the IR is that we lose the `nonnull` attribute.